mbedtls/src/ssl/context.rs

```diff
@@ -158,70 +181,37 @@ impl<'config> Context<'config> {
     }
 
     pub fn verify_result(&self) -> StdResult<(), VerifyError> {
-        match unsafe { ssl_get_verify_result(&self.inner) } {
+        match unsafe { ssl_get_verify_result(self.into()) } {
             0 => Ok(()),
             flags => Err(VerifyError::from_bits_truncate(flags)),
         }
     }
 
-    pub fn config(&self) -> &'config Config {
-        unsafe { UnsafeFrom::from(self.inner.conf).expect("not null") }
-    }
-}
-
-impl<'ctx> HandshakeContext<'ctx> {
-    pub fn set_authmode(&mut self, am: AuthMode) {
-        unsafe { ssl_set_hs_authmode(self.inner, am.into()) }
+    pub fn config(&self) -> &Option<Arc<Config>> {
+        &self.config
     }
-
-    pub fn set_ca_list<C: Into<&'ctx mut LinkedCertificate>>(
-        &mut self,
-        list: Option<C>,
-        crl: Option<&'ctx mut Crl>,
-    ) {
+    
+    #[allow(dead_code)]
+    pub fn close(&mut self) {
         unsafe {
-            ssl_set_hs_ca_chain(
-                self.inner,
-                list.map(Into::into)
-                    .map(Into::into)
-                    .unwrap_or(::core::ptr::null_mut()),
-                crl.map(Into::into).unwrap_or(::core::ptr::null_mut()),
-            )
+            ssl_close_notify(self.into());
+            ssl_set_bio(self.into(), ::core::ptr::null_mut(), None, None, None);
+            self.io = None;
         }
     }
 
-    /// If this is never called, will use the set of private keys and
-    /// certificates configured in the `Config` associated with this `Context`.
-    /// If this is called at least once, all those are ignored and the set
-    /// specified using this function is used.
-    pub fn push_cert<C: Into<&'ctx mut LinkedCertificate>>(
-        &mut self,
-        chain: C,
-        key: &'ctx mut crate::pk::Pk,
-    ) -> Result<()> {
-        unsafe {
-            ssl_set_hs_own_cert(self.inner, chain.into().into(), key.into())
-                .into_result()
-                .map(|_| ())
+    pub fn get_mut_io<T: IoCallback + 'static>(&mut self) -> IoResult<&mut T> {
+        match self.io.as_mut() {
+            Some(stream) => {
+                match stream.as_any().downcast_mut::<T>() {
+                    Some(s) => Ok(s),
+                    None => Err(IoError::new(IoErrorKind::NotFound, "Peer kind is not expected")),
+                }
+            }
+            None => Err(IoError::new(IoErrorKind::NotFound, "No peer available")),
         }
     }
-}
-
-impl<'ctx> ::core::ops::Deref for HandshakeContext<'ctx> {
-    type Target = Context<'ctx>;
-
-    fn deref(&self) -> &Context<'ctx> {
-        unsafe { UnsafeFrom::from(&*self.inner as *const _).expect("not null") }
-    }
-}
-
-impl<'ctx> UnsafeFrom<*mut ssl_context> for HandshakeContext<'ctx> {
-    unsafe fn from(ctx: *mut ssl_context) -> Option<HandshakeContext<'ctx>> {
-        ctx.as_mut().map(|ctx| HandshakeContext { inner: ctx })
-    }
-}
-
-impl<'a> Session<'a> {
```

Comment:
I prefer having the type-level separation. Here's one way to do it without having to duplicate the type:
```rust
pub enum Fresh {}
pub enum Established {}

pub struct Context<S = Fresh> { ... }
pub type Session = Context<Established>;

impl<S> Context<S> {
    pub fn new() -> Context { ... }
    pub fn establish(self) -> Result<Session, (Self, Error, Option<VerifyError>)> { ... }
}

impl Session {
    // functions that only make sense when established
}
```

You can go a step further and also have a `enum Handshake` and use that instead of the separate HandshakeContext type.

Reply:
Why this may be difficult to use / understand:

- We cannot have multiple 'Session' per Context.

- Any new session overrides/closes the old one in the Context.

- Context may manipulate Session indirectly.

As such - I just put everything in Context to avoid any logic issues.

I could make the Session a reference to Context with lifetime but I wanted to avoid lifetimes...

I also wanted to avoid making Context generic due to reasons in another comment above.

Reply:
Parking for now until we get native-tls compat running